test/py/test_migration.py

```diff
@@ -28,139 +29,504 @@
-    """
-    Tests writing to the migration state register, the device not being able to
-    immediately quiesce, and then finally the device failing to transition to
-    the new migration state.
-    """
 
-    global ctx, sock
+def test_handle_mig_data_read():
+    global read_data
 
-    data = VFIO_DEVICE_STATE_V1_RUNNING.to_bytes(c.sizeof(c.c_int), 'little')
-    write_region(ctx, sock, VFU_PCI_DEV_MIGR_REGION_IDX, offset=0,
-                 count=len(data), data=data, rsp=False,
-                 busy=True)
+    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
+
+    argsz = len(vfio_user_mig_data()) + 4
+
+    payload = vfio_user_mig_data(
+        argsz=argsz,
+        size=4
+    )
+
+    data = bytes([0, 1, 2, 3])
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY)
+    read_data = data
+    result = msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload)
+    assert len(result) == argsz
+    assert result[len(vfio_user_mig_data()):] == data
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_STOP_COPY)
+    read_data = data
+    result = msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload)
+    assert len(result) == argsz
+    assert result[len(vfio_user_mig_data()):] == data
+
+
+def test_handle_mig_data_read_too_long():
+    global read_data
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
+
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 8,
+        size=8
+    )
+
+    # When we set up the tests at the top of this file we specify that the max
+    # data transfer size is 4 bytes. Here we test to check that a transfer of 8
+    # bytes fails.
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY)
+    read_data = bytes([1, 2, 3, 4, 5, 6, 7, 8])
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+
+
+def test_handle_mig_data_read_invalid_state():
+    global read_data
+
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 4,
+        size=4
+    )
+
+    data = bytes([1, 2, 3, 4])
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
+    read_data = data
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_STOP)
+    read_data = data
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
 
-    ret = vfu_device_quiesced(ctx, 0)
-    assert ret == 0
 
-    print("waiting for reply")
-    get_reply(sock, errno.ENOTTY)
-    print("received reply")
+def test_handle_mig_data_read_failed_callback():
+    global fail_callbacks
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY)
+
+    fail_callbacks = True
+
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 4,
+        size=4
+    )
+
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+
+    fail_callbacks = False
+
+
+def test_handle_mig_data_read_short_write():
+    payload = struct.pack("I", 8)
```

Comment:
Can you explain a bit more?

Reply:
I'm not sure why I did it like that, I'll explicitly make it send a `vfio_user_mig_data` payload with the data missing.

Reply:
Upon further investigation it has arisen that there is an actual bug and it will just start giving out-of-bounds data to the write callback 😳 I will fix that in the implementation PR.